ui/apps/platform/src/Containers/Clusters/Components/AutoUpgradeToggle.tsx

```diff
@@ -25,11 +26,16 @@ function AutoUpgradeToggle(): ReactElement {
     }, []);
 
     function toggleAutoUpgrade(): void {
+        if (!autoUpgradeConfig) {
+            return;
+        }
+
         // @TODO, wrap this settings change in a confirmation prompt of some sort
         const previousValue = autoUpgradeConfig.enableAutoUpgrade;
         const newConfig = {
             ...autoUpgradeConfig,
             enableAutoUpgrade: !previousValue,
+            autoUpgradeAllowed: Upgradability.NOT_ALLOWED,
```

Comment:
@fredrb This is reason for the problem that Misha reports.

Reply:
The problem is that this won't allow for `undefined` if the property definition isn't defined as `autoUpgradeAllowed?`. Should I rollback the other suggestion and allow the `autoUpgradeConfig` to have this as `undefined`?

Reply:
For now, replace with `Upgradability.ALLOWED` because a prerequisite to render the toggle switch, true?

After a meeting, I will take a look.

Reply:
Aha, notice the early return you added at the beginning of this function?

```js
        if (!autoUpgradeConfig) {
            return;
        }
```

TypeScript error because corresponding `handleChange` function in the PF file needs it too.

Bonus follow up for me is roll back looks incorrect in that function.

Reply:
Three strikes and I am out! JavaScript hoists `function` to top of scope above the early return.

But `const` assignment with arrow function is where it is, so early return does indeed imply to TypeScript that `autoUpgradeConfig` is not null.

It was an intentional decision to delete TODO comment.

```ts
    if (!autoUpgradeConfig) {
        return <></>;
    }

    const toggleAutoUpgrade = () => {
        const previousValue = autoUpgradeConfig.enableAutoUpgrade;
        const newConfig = {
            ...autoUpgradeConfig,
            enableAutoUpgrade: !previousValue,
        };

        setAutoUpgradeConfig(newConfig); // optimistically set value before API call

        saveAutoUpgradeConfig(newConfig).catch(() => {
            // reverse the optimistic update of the control in the UI
            const rollbackConfig = {
                ...autoUpgradeConfig,
                enableAutoUpgrade: previousValue,
            };
            setAutoUpgradeConfig(rollbackConfig);

            // also, re-fetch the data from the server, just in case it did update but we didn't get the network response
            fetchConfig();
        });
    };
```

Change for the PF component. I added comment on line where I fixed mistake in original code.

```ts
    if (!autoUpgradeConfig) {
        return <></>;
    }

    const handleChange = (value: boolean) => {
        setIsDisabled(true);
        const newConfig = {
            ...autoUpgradeConfig,
            enableAutoUpgrade: value,
        };

        setAutoUpgradeConfig(newConfig); // optimistically set value before API call

        saveAutoUpgradeConfig(newConfig)
            .then(() => {
                setIsDisabled(false);
            })
            .catch(() => {
                // reverse the optimistic update of the control in the UI
                const rollbackConfig = {
                    ...autoUpgradeConfig,
                    enableAutoUpgrade: !value, // previous value is opposite to new value
                };
                setAutoUpgradeConfig(rollbackConfig);

                // also, re-fetch the data from the server, just in case it did update but we didn't get the network response
                fetchConfig();
            });
    };
```

You have my apology on behalf of UI team that the code was not firm foundation for you to build on.

Because so many changes, I will write draft integration tests now as a follow up for after you merge.

Reply:
Done in pair